File: solutions/julia/factory-sensors/1/factory-sensors.jl
function humiditycheck(pct_humidity)
    
    ex = ErrorException("humidity level check failed: $pct_humidity%")
    
    if pct_humidity >= 80 
        throw(ex)
    end

    @info "humidity level check passed: $pct_humidity%"

end

temperaturecheck(temperature::Nothing) = throw(ArgumentError("sensor is broken"))
temperaturecheck(temperature::Int64) =
    temperature > 500 ?
    throw(DomainError(temperature, "overheating detected: $temperature °C")) :
    @info "temperature check passed: $temperature °C"
    
struct MachineError <: Exception end

function machinemonitor(pct_humidity, temperature)
    failed = false

    try
        humiditycheck(pct_humidity)
    catch problem
        if problem isa ErrorException
            @error "humidity level check failed: $pct_humidity%"
            failed = true
        else
            rethrow()
        end
    end

    try
        temperaturecheck(temperature)
    catch problem
        if problem isa ArgumentError
            @warn "sensor is broken"
            failed = true
        elseif problem isa DomainError
            @error "overheating detected: $temperature °C"
            failed = true
        else
            rethrow()
        end
    end

    if failed
        throw(MachineError())
    end


end

File: solutions/r/factory-sensors/1/factory-sensors.R
check_humidity_level <- function(pct_humidity) {
  stopifnot(pct_humidity <= 70)
  message("humidity test passed")
  TRUE
}

report_overheating <- function(temperature) {
  if(is.null(temperature)) {
    stop("Sensor Broken")
  } else if(temperature > 600) {
    stop("Overheating: ", temperature, "°C") 
  } else if(temperature > 500) {
    warning("Risk of overheating: ", temperature, "°C")
    return(F)
  }
  message("Temperature check passed: ", temperature, "°C")
  return(T)
}

monitor_the_machine <- function(pct_humidity, temperature) {
  c <- suppressWarnings(try(check_humidity_level(pct_humidity)))
  r <- suppressWarnings(try(report_overheating(temperature)))
  if(c && r) return(message("All OK"))
}
